ocaml/typing/ctype.ml

```diff
@@ -2521,18 +2521,23 @@ let unexpanded_diff ~got ~expected =
 
 (**** Unification ****)
 
-(* Return whether [t0] occurs in [ty]. Objects are also traversed. *)
-let deep_occur t0 ty =
+(* Return whether [t0] occurs in any type in [tyl]. Objects are also traversed. *)
+let deep_occur_list t0 tyl =
   let rec occur_rec ty =
     if get_level ty >= get_level t0 && try_mark_node ty then begin
       if eq_type ty t0 then raise Occur;
       iter_type_expr occur_rec ty
     end
   in
   try
-    occur_rec ty; unmark_type ty; false
+    List.iter occur_rec tyl;
+    List.iter unmark_type tyl;
+    false
   with Occur ->
-    unmark_type ty; true
+    List.iter unmark_type tyl;
+    true
+
+let deep_occur t0 ty = deep_occur_list t0 [ty]
```

Comment:
One thing that's a tiny bit sad is we're now allocating a list cell in the common case that we do just have one type.  Maybe it would be worth pulling the core `occur_rec` functionality out and having both `deep_occur_list` and `deep_occur` call it directly, rather than defining `deep_occur` in terms of `deep_occur_list`?

Reply:
Makes sense, updated!